Add VP9 video codec support

himmelroman opened the pull request:
First off - I LOVE aiortc! I've been using it for years with great success in production. Thank you to all the maintainers and contributors for this amazing library! 🙏

## Summary

This PR adds full VP9 video codec support to aiortc, following RFC 9628 (RTP Payload Format for VP9 Video). The implementation is an exact copy of Pion WebRTC's VP9 implementation in Go, ensuring compatibility and correctness.

## Features

- ✅ Full RFC 9628 compliance for parsing (all modes supported)
- ✅ Both flexible mode (F=1) and non-flexible mode (F=0) encoding
- ✅ VP9 header parser for bitstream analysis
- ✅ Scalability Structure (V=1) support on keyframes
- ✅ Browser-tested with Chrome (both modes working)
- ✅ Configurable via `flexible_mode` parameter (defaults to True for maximum browser compatibility)

## Implementation Details

### Core Components

**Vp9PayloadDescriptor:**
- Complete RFC 9628 payload descriptor with full serialization/parsing
- Supports 7-bit and 15-bit Picture IDs
- Layer indices (TID, SID, U, D flags)
- Reference indices (P_DIFF) for flexible mode
- TL0PICIDX for non-flexible mode
- Scalability Structure parsing (V=1)

**Vp9Encoder:**
- Uses PyAV's libvpx-vp9 with WebRTC-optimized settings
- Configurable flexible/non-flexible mode
- Picture ID and TL0PICIDX management
- Bitrate control and multi-threaded encoding

**Vp9Decoder:**
- libvpx-vp9 decoder integration
- RTP depacketization
- Jitter buffer compatibility

**VP9 Header Parser:**
- `_parse_vp9_header()`: Bitstream parser for keyframe/inter-frame detection
- Extracts frame dimensions from keyframes
- EXACT copy of Pion's header.go logic

**Packetization Methods:**
- `_packetize_flexible()`: EXACT copy of Pion's payloadFlexible()
- `_packetize_non_flexible()`: EXACT copy of Pion's payloadNonFlexible()

## Testing

- ✅ 25 new VP9 unit tests (13 descriptor + 12 encoder/decoder)
- ✅ 1 integration test in test_rtcpeerconnection.py
- ✅ All 519 aiortc tests passing
- ✅ Browser compatibility verified with Chrome (both flexible and non-flexible modes)

## Files Modified

- `src/aiortc/codecs/vpx.py` (+~700 lines) - VP9 implementation
- `src/aiortc/codecs/__init__.py` (+18 lines) - Codec registration
- `tests/test_vpx.py` (+~370 lines) - Unit tests
- `tests/test_rtcpeerconnection.py` (+~100 lines) - Integration test
- `tests/test_rtcrtpreceiver.py` (+2 lines) - Updated capabilities
- `tests/test_rtcrtpsender.py` (+2 lines) - Updated capabilities
- `examples/server/index.html` (+1 line) - VP9 codec option
- `src/aiortc/contrib/media.py` (+4 lines) - MediaRecorder VP9 support
- `README.rst` (+1 line) - Updated supported codecs list

## Reference Implementation

This implementation is an EXACT copy of Pion WebRTC's VP9 implementation to ensure correctness:
- [Pion RTP VP9 Packet](https://github.com/pion/rtp/blob/master/codecs/vp9_packet.go)
- [Pion VP9 Header Parser](https://github.com/pion/rtp/blob/master/codecs/vp9/header.go)
- [Pion VP9 Bit Reading](https://github.com/pion/rtp/blob/master/codecs/vp9/bits.go)
- [RFC 9628: RTP Payload Format for VP9 Video](https://datatracker.ietf.org/doc/html/rfc9628)

## Backwards Compatibility

✅ Yes - This PR only adds new codec support without breaking any existing functionality.

## Notes

- VP9 provides ~30% better compression than VP8 at similar quality
- Default flexible mode (F=1) ensures maximum browser compatibility
- Non-flexible mode (F=0) available for specific use cases requiring TL0PICIDX

---

## UPDATE: Payload Type Collision Fix (60bab9b)

Addressed a codec negotiation bug reported by @Arlaz where PT 103 was assigned to both H264 RTX and VP9 when no codec preferences were set, causing browsers to reject the SDP with "Duplicate payload type with conflicting codec name or clock rate".

**Fix:** Ported libwebrtc's PT collision resolution algorithm to `find_common_codecs()`:
- **Phase 1:** Prefer remote's PT if available (minimizes SDP changes)
- **Phase 2:** Search descending from 127 when collisions occur

**Changes:**
- Refactored `rtcpeerconnection.py` with `find_unused_id()` and `find_and_set_id_used()` helpers
- Added BUNDLE support via `used_payload_types` parameter for session-wide PT namespace
- Enhanced RTX codec handling to properly update `apt` references

**Tests:** Added 3 new test cases covering PT collision scenarios and BUNDLE mode.

Arlaz commented:
I did some test using this PR on a a bare aiortc project and VP9 addition is generating a broken SDP answer where the same payload type (PT 103) is used for two different codecs (VP9 and H264). Browsers reject that with this error :

“[Duplicate payload type with conflicting codec name or clock rate.](Negotiation failed: OperationError: Failed to execute 'setRemoteDescription' on 'RTCPeerConnection': Failed to parse SessionDescription. a=rtpmap:103 VP9/90000 ... a=rtpmap:103 H264/90000 Duplicate  payload type with conflicting codec name or clock rate.)”

himmelroman commented:
@Arlaz Hey! 
I dug into your duplicate PT error but couldn't reproduce it. All my tests show unique PTs for each codec. 
Questions: 
1. Which browser did you test with? 
2. Could you share the actual **test code** and **SDP** that shows the duplicate PT 103?

While trying to reproduce your error, I found myself cleaning up the codec forcing approach.
I had some tests manually stripping codecs from SDP - which was ugly and hacky. 
Now everything is using proper setCodecPreferences() API:

```
# get sender & transceiver
sender = pc.addTrack(video_track)
transceiver = next(t for t in pc.getTransceivers() if t.sender == sender)

# select vp9
vp9_codecs = [c for c in RTCRtpSender.getCapabilities("video").codecs if c.mimeType == "video/VP9"]

# set preference
transceiver.setCodecPreferences(vp9_codecs)

# then normal offer/answer flow
offer = await pc.createOffer()
```

The webcam example already does it this way if you run it with `--video-codec video/VP9`. 
If you can share your test code and the SDP showing duplicate PTs - I will gladly look into it and fix whatever causes this!

Thanks!

jlaine commented on src/aiortc/codecs/vpx.py:
```diff
@@ -166,6 +167,325 @@ def parse(cls: Type[DESCRIPTOR_T], data: bytes) -> tuple[DESCRIPTOR_T, bytes]:
+        self.E = end_of_frame
+        self.V = scalability_structure_present
+        self.Z = not_reference_frame
+
```
I much prefer the arguments you're passing to the constructor (more explicit) that the single letter member names. You might also consider just making `Vp9PayloadDescriptor` a `dataclass` to save some boilerplate.

himmelroman replied:
Agreed and done, I prefer them much more myself!
Also refactored to `dataclass` - nice and clean now.

jlaine commented on src/aiortc/codecs/vpx.py:
```diff
@@ -166,6 +167,325 @@ def parse(cls: Type[DESCRIPTOR_T], data: bytes) -> tuple[DESCRIPTOR_T, bytes]:
+        if not self.F and self.tl0picidx is not None:
+            flags.append(f"TL0={self.tl0picidx}")
+
+        return f"Vp9PayloadDescriptor({', '.join(flags)})"
```
This is not exercised by any tests, and if you use a dataclass I suspect it's not needed at all.

himmelroman replied:
Removed as per your suggestion.

himmelroman commented:
@jlaine Addressed your comments and suggestions.
Please review again now.
And also see separate PR for PT Negotiation fix: https://github.com/aiortc/aiortc/pull/1401

himmelroman commented:
@jlaine Are there other maintainers I can ping for this?
